paste: redesign, fixes, 10x speedup

justinmk opened the pull request:
This PR improves paste behavior and performance.

- Performance: 10x-100x faster (try pasting MB of input, or paste a lot of text with `filetype=cpp` or `filetype=ruby`)
- Extensibility: users can control paste behavior by redefining `vim.paste` (Lua)
- API: `nvim_paste` for UIs/clients
- API: `nvim_put` inserts text (weird but true: this was difficult before)
- Fixes:
  - fix #3447
  - fix #3566
  - fix #7066
  - fix #7212
  - fix #7273
  - fix #7455
  - fix #10415

Tasks:

- [x] implement [as callback](https://github.com/neovim/neovim/pull/4448#issuecomment-196424051) instead of pre/post autocmds
- [x] Indicate paste start/stop as parameter to handler?
- [x] generalize `string_to_array()` ?
- [x] API: `nvim_put` #6819
- [x] API: `nvim_paste`
- [x] redo (dot-repeat)
- [x] 'nomodifiable' buffer
- [x] documentation

TODO (future):

- Replace/Operator-pending/Visual modes
- remove 'pastetoggle' `p_pt`
- Dot-repeat is slow (same behavior in Vim) because it is subject to Vim auto-indent/etc.
- Mechanism for  associating a callback with a `nvim_input` chunk. #10826
- `TextPutPre`/`TextPutPost` (analogs for `TextYankPost`) ?
- Send paste sequences to `:terminal`
  - As @ZyX-I [mentioned](https://github.com/neovim/neovim/issues/3476#issuecomment-149519446) this would also be warranted for `p`, `:put`, etc.
- handle missing "stop paste" code (how? timeout...?) #10865

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -3435,11 +3435,11 @@ getchar([expr])						*getchar()*
 <
 		You may also receive synthetic characters, such as
-		|<CursorHold>|. Often you will want to ignore this and get
+		|<LeftMouse>|. Often you will want to ignore this and get
```
`<CursorHold>` pseudokey doesn't exist any more (replaced by K_EVENT).

bfredl replied:
`<LeftMouse>` is user input, just like `<F10>` or any other special key without canonical unicode representation, of which there are many. Seems weird to ignore only it specifically.  CursorHold needed extra treatment here because it was generated by nvim without user intervention so `getchar()` terminated early, other special chars do not have this problem.

justinmk replied:
Good point. Maybe we can remove this blurb from the docs?

bfredl replied:
I think so. The concrete examples above of how to actually _handle_ pseudokeys (including mouse) should be enough.

justinmk commented on src/nvim/getchar.c:
```diff
@@ -1868,14 +1866,8 @@ static int vgetorpeek(int advance)
-              // didn't match ui_toggle_key and didn't try the whole typebuf,
-              // check the 'pastetoggle'
-              match = typebuf_match_len(p_pt, &mlen);
-            }
```
Now we don't need this special-case logic in this low-level function...

justinmk commented on src/nvim/keymap.c:
```diff
@@ -282,7 +282,7 @@ static struct key_name_entry {
   {K_SNR,             (char_u *)"SNR"},
   {K_PLUG,            (char_u *)"Plug"},
-  {K_PASTE,           (char_u *)"Paste"},
+  {K_PASTEPOST,       (char_u *)"PastePost"},
```
See `tui/input.c` for why we don't need a "PastePre" pseudo-key. I tried to avoid "PastePost", but couldn't (and have some questions about that).

justinmk commented on src/nvim/tui/input.c:
```diff
@@ -233,9 +237,12 @@ static void tk_getkeys(TermInput *input, bool force)
   }
 
-  if (result != TERMKEY_RES_AGAIN || input->paste_enabled) {
+  if (result != TERMKEY_RES_AGAIN /*|| input->paste_started */) {
```
@tarruda Is this skip just an optimization?

tarruda replied:
No, but this probably isn't needed anymore(I don't think it is possible to have `result == TERMKEY_RES_AGAIN && input->paste_enabled` because `read_cb` will only push data up to the next `\x1b`(esc). I'm not sure why this is there, probably an oversight or leftover from previous code.

justinmk commented on src/nvim/tui/input.c:
```diff
@@ -282,27 +289,56 @@ static bool handle_focus_event(TermInput *input)
+    if (enable) {
+      loop_schedule(&loop, event_create(1, apply_pastepre, 0));
+    } else {
+      enqueue_input(input, PASTEPOST_KEY, sizeof(PASTEPOST_KEY) - 1);
```
@tarruda / @oni-link, Wondering if you can help my understanding. `loop_schedule()` on the "start paste" case (line 306) works well, because it puts the event on the main queue before any input gets processed. But for the "stop paste" case, I had to enqueue this `PASTEPOST_KEY` in order to guarantee that all paste data was processed before queuing the `PastePost` event. 

 Do we currently have any better mechanism for placing an event at a precise location in the input stream (I'm guessing the answer is "no", hence why we still have these pseudokeys)? I also tried using a mutex, but that had problems and isn't really desirable anyways.

I am wondering if it would be a good idea to provide some internal generic mechanism for placing `K_EVENT` in the input stream and associating it with a callback.

tarruda replied:
> Do we currently have any better mechanism for placing an event at a precise location in the input stream (I'm guessing the answer is "no", hence why we still have these pseudokeys)

No we don't.

> I am wondering if it would be a good idea to provide some internal generic mechanism for placing K_EVENT in the input stream and associating it with a callback.

K_EVENT is not meant to be used by UIs, it is only a general mechanism for waking nvim when an asynchronous event happens. An UI placing a bunch of keys in the input stream is an asynchronous event, but not "certain key gets processed"(And the only use case I could think for this is toggling paste, so I'm not sure it would be useful)

justinmk replied:
> "certain key gets processed"(And the only use case I could think for this is toggling paste, so I'm not sure it would be useful)

@tarruda So does that mean we can safely use `loop_schedule()` to schedule FocusGained/FocusLost instead of the K_FOCUSGAINED/K_FOCUSLOST?

tarruda replied:
> @tarruda So does that mean we can safely use loop_schedule() to schedule FocusGained/FocusLost instead of the K_FOCUSGAINED/K_FOCUSLOST?

Possibly yes. I may be missing something, but FocusGained/FocusLost doesn't seem to be events that needs to be handled at an exact point during input processing.

justinmk commented on src/nvim/tui/input.c:
```diff
@@ -376,8 +412,8 @@ static void read_cb(Stream *stream, RBuffer *buf, size_t c, void *data,
-  // without wrap around, otherwise it could be misinterpreted.
+  // CSI and K_SPECIAL are double-escaped later (input_enqueue).
+  // Make sure they can fit into the ring buffer without wrap around.
   rbuffer_reset(input->read_stream.buffer);
```
@tarruda @oni-link I tried to clarify this a bit. Is the updated comment accurate?

tarruda replied:
No, `input->read_stream.buffer` is only read by the loop above. There could be an multi-byte key sequence that is trimmed by the end of the buffer(wrapping around to the beginning), but `termkey_push_bytes` is not aware of that, so it could misinterpret the key. `rbuffer_reset` simply moves the data so it is all in contiguous memory.

justinmk replied:
Oh, I get it: this problem occurs because of `rbuffer_read_ptr()` exposing the raw buffer.

justinmk commented on test/functional/tui/child_session.lua:
```diff
@@ -1,3 +1,9 @@
+-- jobsend() to send raw bytes. This allows us to test the tui/input.c module,
+-- which does have a public API. But this means that the test session methods
+-- (eval(), request(), ... from helpers.lua) operate on the _host_ session,
+-- _not_ the embedded session.
```
This is very confusing for people trying to follow the tests. I hope to improve this some more, but for now I renamed `terminal/helpers.lua` to `tui/child_session.lua`.

tarruda commented:
@justinmk I'm not sure I understand the goal behind this PR, is it to decouple the logic of automatically toggling `paste` with a special key?

If so, I don't see much gain here: We still have bracketed paste parsing hardcoded into the TUI, only the `set paste` logic will be moved to vimscript(even so, what would be the purpose of allowing autocmds to be attached to paste?). But I do see a worthly goal in this:

> I am wondering if it would be a good idea to provide some internal generic mechanism for placing K_EVENT in the input stream and associating it with a callback.

If you could implement this, then yes, we could completely decouple bracketed paste(and possibly many other terminal sequences) from the tui. But it should not be `K_EVENT`, since it is already used for another purpose(and no, we can't place arbitrary events so they are processed at specific times during input processing).

However, this seems to overlap with @bfredl's work on #4419 :

```vim
map <command> <Paste> if &paste | set nopaste | else | set paste
```

Maybe it would be even better if we simply supported user-defined synthetic keys:

```vim
map <command> <User:SOME_KEY_ID> if &paste | set nopaste | else | set paste
```

The question is: How can a plugin tell the UI to put such synthetic keys into the input buffer? Right now we can't, but the TUI will soon allow configuration, so a bracketed paste plugin could be summarized to something like this:

```vim
let g:tui_keys_override = {
\  '\x1b[200~': '<User:Paste>'
\  '\x1b[201~': '<User:Paste>'
\}
imap <command> <User:Paste> if &paste | set nopaste | else | set paste
``` 

That is, the plugin asks the tui to translate bracketed pastes into `<user:paste>`, and maps it to a code that toggles the paste option.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -328,6 +326,9 @@ int main(int argc, char **argv)
 #undef PROTO
 
+  do_cmdline_cmd("autocmd PastePre  * set paste|if mode()!~#'c\\|t'|call feedkeys('\x1C\xEi','n')|endif");
+  do_cmdline_cmd("autocmd PastePost * set nopaste|if mode()!~#'c\\|t'|call feedkeys('\x1C\xE','n')|endif");
```
`\xE` looks weird, better write `\x0E`.

ZyX-I replied:
Though no, it is better to write this as `\"\\<C-\\>\\<C-n>\"`.

ZyX-I replied:
And I would also suggest to create autocmd group `_Builtin` for all built-in autocommands (currently only this one and term://\* BufReadCmd), so that they are easy to identify, making user aware of what he sees and, possibly, changes.

ZyX-I replied:
Regex is better written as `stridx('ct', mode()) == -1` or `mode() !~# '[ct]'`.

justinmk commented:
> If so, I don't see much gain here:

This fixes several bugs, and removes some special-case logic.

> We still have bracketed paste parsing hardcoded into the TUI, only the set paste logic will be moved to vimscript

Hard-coded handling of bracketed paste _belongs_ in `tui/input.c`--that's not a problem. And this PR  removes inspection of `ui_toggle[]` from `vgetorpeek()`. That's a net gain.

> (even so, what would be the purpose of allowing autocmds to be attached to paste?). 

I addressed that here: https://github.com/neovim/neovim/pull/3906#issuecomment-168349576 I'll also add that some users would want to control whether they return to normal-mode after a paste, or whether a newline is always created.

Moreover, @bfredl mentioned wanted `Paste` event as the dual for `TextYankPost`.

> However, this seems to overlap with @bfredl's work on #4419 :

I don't think so. The `<Paste>` toggle doesn't work. The system needs to know "this is the start" and "this is the end". The state of `'paste'` cannot be a factor in this, the user may already have it enabled.

justinmk commented on src/nvim/tui/input.c:
```diff
@@ -282,27 +283,39 @@ static bool handle_focus_event(TermInput *input)
+
+    if (enable && input->paste_started) {
+      // Received a bogus "paste start" after paste was already started.
+      return false;
```
Old code didn't handle this case.

bfredl commented:
`TextPutPre`(/`Post` ?) and `PastePre`/`PastePost` seems to me to be completely different events.The former would handle putting,  `p` (and maybe `i<c-r>` and `:<c-r>` as well) while the event here is for terminal bracketed paste.

justinmk commented:
@bfredl See the PR description where I mention that blackhole register could be used to mean "external paste" (called "bracketed paste" in terminals, but for all other UIs it's a "clipboard paste").

bfredl commented:
called "bracketed paste" in terminals, but for all other UIs it's a "clipboard paste"

Not really, `"+p` or `"*p` (or the insert mode variants) would be the closest nvim notion of a UI clipboard paste. The difference is that terminal paste (at least before this PR?) doesn't do any putting operation, rather it changes behavior of insert mode to insert typed text verbatim. (but it doesn't for instance switch to insert mode automatically)

justinmk commented:
>  The difference is that terminal paste (at least before this PR?) doesn't do any putting operation, rather it changes behavior of insert mode to insert typed text verbatim

GUI (gvim) has similar operation for middle-click, cmd-v, etc. Call it whatever you like: "system paste", "platform paste", "external paste".

ZyX-I commented:
@justinmk Bracketed paste is terminal-UI-specific thing. GUIs do not have anything like this, all they have are events from user: clipboard and mouse. There is no “external paste”, there are things like `"+p`, `:put` and events which map to this. In any case clipboard contents is received in one piece, it is not a stream like `{paste start}{contents}{paste end}`.

ZyX-I commented:
@justinmk `<MiddleMouse>`, `<C-v>`, etc are handled by the application *itself* and it is UI which initiates pasting. There are *conventions* and toolkits which follow conventions by default, but they are not enforced and toolkits can be told to not follow them as well. With bracketed paste it is terminal which initiates pasting, which is an “external” application.

justinmk commented:
> `<MiddleMouse>`, `<C-v>`, etc are handled by the application itself and it is UI which initiates pasting.

Yes, and `tui/input.c` _is_ the application layer in the case of the TUI. In the case of an external UI, analogously, it would raise `PastePre` and `PastePost` and send the bytes in between.

I don't understand the un-ending pedantic notes in this PR.

bfredl commented:
That two different operation sometimes can have the the same goal does not make them the same operation, I don't see any pedantry in that.
In `TextYankPre` the autocmd could be allowed to change the text and the register mode. What would that even mean for `PastePre`. (The oprational sematics are the relevant here and not any naming)

ZyX-I commented:
@justinmk It is only TUI which *has* to “raise PastePre, send bytes, PastePost” (and “has” is mostly because Neovim needs to be prepared to missing PastePost). Other UIs may say “paste this”, but they don’t need this *at all*, there is no “external” paste, *nobody* will raise PastePre, PastePost and send bytes because pasting is done by `"*p` or equivalents. The only thing is that some UIs may want to alter clipboard providers.

When “Paste\*” is used for handling pastes from register `a` semantics changes significantly: previously input was put into an input queue. Now input is located in register `a` and nobody wants it in the input queue.

justinmk commented:
> In TextYankPre the autocmd could be allowed to change the text and the register mode. What would that even mean for PastePre

Not sure I follow. We increment `textlock` for `TextYankPost`. We can set textlock for Paste events as well, regardless of what we do for `TextYankPre`.

> Other UIs may say “paste this”, but they don’t need this at all, there is no “external” paste, nobody will raise PastePre, PastePost and send bytes because pasting is done by "*p or equivalents.

Why couldn't we recommend that UIs raise the events? This would enable users to have consistent paste behavior for any UI. E.g., some users may want to return to normal-mode _always_ after a paste. Some users may want to respond to a non-modifiable buffer by opening a new tab instead of just showing an error.

> When “Paste*” is used for handling pastes from register a semantics changes significantly: previously input was put into an input queue. Now input is located in register a and nobody wants it in the input queue.

Fair point, then we can drop that idea. It's not part of this PR.

bfredl commented:
Sorry meant TextPutPre

ZyX-I commented:
By the way, consider that user pressed `<S-Insert>foo`, which translated into `{paste start}{inserted contents}{paste end}foo`. Can I use Paste\* events to get `{inserted contents}` without `foo`?

Also I saw that you test ignoring second `{paste start}` in `{paste start}{paste start}`. This is incorrect, I checked konsole using

    echo $'\e[200~' | xclip -i
    echo $'\e[?2004h' ; stty -echo ; cat -v

(with `<MiddleMouse>` after last `<CR>`) and got

    ^[[200~^[[200~
    ^[[201~

: no escaping. Spurious `{paste start}` should be part of the pasted text, not ignored. Double `{paste end}`/`{paste end}` without `{paste start}` should be ignored though I think.

justinmk commented:
> Can I use Paste* events to get {inserted contents} without foo?

The text is not provided in any register or `v:` variable, but the buffer contents at time of `PastePost` should not contain `foo`. And you could compare this to buffer contents at time of `PastePre`.

>  Spurious {paste start} should be part of the pasted text, not ignored. Double {paste end}/{paste end} without {paste start} should be ignored though I think.

Thanks, I will fix it.

tarruda commented:
> I don't think so. The <Paste> toggle does not work. It's broken. The system needs to know "this is the start" and "this is the end". The state of 'paste' cannot be a factor in this, the user may already have it enabled.

> The old code was sending <Paste> to toggle on and another one later in the stream to toggle off. This is the cause of bugs we've had for 6 months.

Alright. What about your [suggestion](https://github.com/neovim/neovim/pull/4448#discussion_r55960175) of adding a generic mechanism to allow execution of vimscript when a certain synthetic keys are processed? Assuming the TUI allows the user to override terminal code parsing(it will), a bracketed paste plugin could do this:

```vim
let g:tui_keys_override = {
\  '\x1b[200~': '<User:PasteEnable>'
\  '\x1b[201~': '<User:PasteDisable>'
\}
```

It seems like something that could work. Even FocusGain/FocusLost could work nicely on such infrastructure.

ZyX-I commented:
I also do not see tests for pasting while in all other modes (visual, select, operator-pending, replace, virtual replace).

tarruda commented:
Adapting my previous comment to @bfredl's [idea](https://github.com/neovim/neovim/pull/4419#issuecomment-196271318), here's how bracketed paste would be implemented as a plugin:

```vim
let g:tui_keys_override = {
\  '\x1b[200~': '<Cmd>set paste<cr>'
\  '\x1b[201~': '<Cmd>set nopaste<cr>'
\}
```

justinmk commented:
> Assuming the TUI allows the user to override terminal code parsing(it will), a bracketed paste plugin could do this:
> let g:tui_keys_override

I suppose that would be fine if no one is convinced that PastePre/PastePost will be useful for GUIs (I gave examples above). But:

- It will block this fix for a long time. Paste has been broken for more than 6 months.
- It means we're back to asking users to fiddle with termcodes--even when the work was already done to abstract it for certain cases. 
- It won't apply to GUIs, so configuration of this behavior won't be consistent across UIs.

bfredl commented:
Hmm I wonder how to implement "Send paste sequences to :terminal" with that. Is the remapping done by the tui module or do you mean runtime will add those mappings to insert mode?

ZyX-I commented:
@bfredl Neither of suggested variants allow implementing “send paste sequences to `:terminal`”, because this needs hooking `"*p`. Though bracket pasting to terminal may be implemented based on PastePre/PastePost, this also needs getting terminal state.

bfredl commented:
> (I gave examples above)

The examples you gave above for gvim are using regisers AFAIK. But it could still make sense for a GUI to do bracketed paste (it could make sense to paste commands in normal mode, which is not the same as `@*` )

bfredl commented:
@ZyX-I I meant specifically @tarruda:s remark.

ZyX-I commented:
@justinmk I want to capture incoming text, not add it anywhere. Actually for handling specifically bracketed paste it would be more useful to have callback that receives input, something like

```VimL
function PasteCallback(text, ended)
  if mode() is# 'n'
    call feedkeys("\<C-\>\<C-n>", 'n')
  endif
  set paste
  let text = a:text[:]
  while !empty(text)
    let line = remove(text, 0)
    let line = substitute(text, "\n", "\<C-v>\<Nul>", 'g')
    let line = substitute(text, "[\x01-\x1F]", "\<C-v>\\0", 'g')
    call feedkeys(line, 'n')
    if !empty(text)
      call feedkeys("\n", 'n')
    endif
    if stridx(line, "\<C-c>")
      return 1
    endif
  endwhile
  if a:ended
    set nopaste
  endif
  return 0
endfunction
```

This means that callback will be invoked *in place* of feeding characters into an input queue. Callback may prevent itself from being called by returning `1`.

Note that I handle special characters. I did not see `<Esc>` pasting example in your tests, but I guess this will interrupt the paste.

ZyX-I commented:
@bfredl *Any* solution which allows calling commands may fix the problem, with the remark that one needs to export terminal state (i.e. whether terminal allows bracketed paste or not). Just create a command which feeds required escape sequences.

tarruda commented:
> Hmm I wonder how to implement "Send paste sequences to :terminal" with that. Is the remapping done by the tui module or do you mean runtime will add those mappings to insert mode?

The `let g:tui_keys_override = {` line is how I envision that UIs will be able to create arbitrary configuration parameters that can be set in a nvimrc or a plugin through [dictwatcheradd()](https://neovim.io/doc/user/eval.html#dictwatcheradd()). In this example, the TUI would have a "keys_override" parameter that allows the user to customize how certain terminal sequences should be translated to keys.

justinmk commented:
> This means that callback will be invoked in place of feeding characters into an input queue.

That will be slower, though. And more complicated to implement.

And, paste has been broken for more than 6 months.

bfredl commented:
Dunno,  I think @justinmk:s solution is better (the only thing I was against was confusing with register putting, in general I like this PR :). This is standard behavior of terminals we should support builtin.

tarruda commented:
> @justinmk I want to capture incoming text, not add it anywhere. Actually for handling specifically bracketed paste it would be more useful to have callback that receives input, something like

@ZyX-I it could be implemented with @bfredl solution and some kind of input redirection mechanism(`:redir` but for user input):

```vim
let g:tui_keys_override = {
\  '\x1b[200~': '<Cmd>:input_redir => g:input<cr>'
\  '\x1b[201~': '<Cmd>:iredir END | call PasteCallback(g:input)<cr>'
\}
```

tarruda commented:
> It will block this fix for a long time. Paste has been broken for more than 6 months.

Good point :+1: 

This generalization can be deferred for when @bfredl finishes #4419

bfredl commented:
Maybe, but I don't see any point of the user ever typing `x1b[200~` in `vimrc`. This should just be a recognized key or autocommand, as this PR does.

ZyX-I commented:
@tarruda Input redirection is one of the worst ideas because it will capture all input, not allowing user to do *anything* until redirection ends.

Also how you are going to stop input redirection if `<Cmd>…` is feed into the input queue and thus will be consumed by input redirection?

tarruda commented:
> @tarruda Input redirection is one of the worst ideas because it will capture all input, not allowing user to do anything until redirection ends.

Good point

> Also how you are going to stop input redirection if <Cmd>… is feed into the input queue and thus will be consumed by input redirection?

Input redirection would operate at a higher level, interpreting some special keys such as `<cmd>` before sending the input to whatever sink it is redirected to(it wouldn't make sense to redirect certain special keys)

ZyX-I commented:
If I am not mistaking one may paste special keys.

@justinmk What #2523 is doing in the list of closed issues if it is not Neovim issue at all?

tarruda commented:
> Maybe, but I don't see any point of the user ever typing x1b[200~ in vimrc. This should just be a recognized key or autocommand, as this PR does.

Sure, it was just an idea of how to decouple this from the TUI. AFAIK the bracketed paste sequence is not standardized, so theoretically a terminal could use a different sequence to implement it(But I agree that something so popular should be supported out of box)

ZyX-I commented:
@justinmk I would not say it is more complicated:

```Patch
diff --git a/src/nvim/tui/input.c b/src/nvim/tui/input.c
index b41e4d2..e900aef 100644
--- a/src/nvim/tui/input.c
+++ b/src/nvim/tui/input.c
@@ -68,17 +68,65 @@ static void input_done_event(void **argv)
   input_done();
 }
 
+static Array string_to_array(const String input)
+{
+  Array ret = { .size = 0, .items = NULL };
+  for (size_t i = 0; i < input.size; i++) {
+    const char *const start = input.data + i;
+    const char *const end = xmemscan(start, NL, input.size - i);
+    i += (size_t) (end - start);
+    ret.size++;
+  }
+  ret.items = xmalloc(ret.size * sizeof(*ret.items));
+  size_t array_idx = 0;
+  for (size_t i = 0; i < input.size; i++) {
+    const char *const start = input.data + i;
+    const size_t line_len
+        = (size_t) ((char *) xmemscan(start, NL, input.size - i)
+                                      - start);
+    i += line_len;
+
+    String item = {
+      .size = line_len,
+      .data = xmemdupz(start, line_len),
+    };
+    memchrsub(item.data, NUL, NL, line_len);
+    ret.items[array_idx++] = STRING_OBJ(item);
+  }
+  ret.capacity = ret.size;
+
+  return ret;
+}
+
 static void wait_input_enqueue(void **argv)
 {
   TermInput *input = argv[0];
   RBUFFER_UNTIL_EMPTY(input->key_buffer, buf, len) {
-    size_t consumed = input_enqueue((String){.data = buf, .size = len});
-    if (consumed) {
-      rbuffer_consumed(input->key_buffer, consumed);
-    }
-    rbuffer_reset(input->key_buffer);
-    if (consumed < len) {
-      break;
+    const String keys = { .data = buf, .size = len };
+    if (input->paste_enabled) {
+      Object keys_array = ARRAY_OBJ(string_to_array(keys));
+      Array args = { .capacity = 1, .size = 1, .items = &keys_array };
+      Error err = ERROR_INIT;
+      Object fret = vim_call_function(STATIC_CSTR_AS_STRING("PasteCallback"),
+                                      args, &err);
+      if ((fret.type == kObjectTypeInteger && fret.data.integer)
+          || (fret.type == kObjectTypeBoolean && fret.data.boolean)
+          || (fret.type == kObjectTypeString && fret.data.string.size)) {
+        input->paste_enabled = false;
+      }
+      api_free_object(fret);
+      api_free_object(keys_array);
+      rbuffer_consumed(input->key_buffer, len);
+      rbuffer_reset(input->key_buffer);
+    } else {
+      const size_t consumed = input_enqueue(keys);
+      if (consumed) {
+        rbuffer_consumed(input->key_buffer, consumed);
+      }
+      rbuffer_reset(input->key_buffer);
+      if (consumed < len) {
+        break;
+      }
     }
   }
   uv_mutex_lock(&input->key_buffer_mutex);
@@ -286,12 +334,18 @@ static bool handle_bracketed_paste(TermInput *input)
       (!rbuffer_cmp(input->read_stream.buffer, "\x1b[200~", 6) ||
        !rbuffer_cmp(input->read_stream.buffer, "\x1b[201~", 6))) {
     bool enable = *rbuffer_get(input->read_stream.buffer, 4) == '0';
+    if (input->paste_enabled && enable) {
+      // Pasting enabling paste literally.
+      return false;
+    }
     // Advance past the sequence
     rbuffer_consumed(input->read_stream.buffer, 6);
     if (input->paste_enabled == enable) {
       return true;
     }
-    enqueue_input(input, PASTETOGGLE_KEY, sizeof(PASTETOGGLE_KEY) - 1);
+    if (!enable) {
+      flush_input(input, true);
+    }
     input->paste_enabled = enable;
     return true;
   }
```

ZyX-I commented:
@justinmk This is, of course, alpha, but I needed to only alter two places. Plus the definition of PasteCallback is needed to be put into main.c, also an indirection (hardcoding function name is not a good idea).

It also appears that I understood the format of incoming string incorrectly, so additional action is needed in string_to_array.

justinmk commented on test/functional/tui/paste_spec.lua:
```diff
@@ -0,0 +1,178 @@
+
+  -- TODO
+  it('handles missing "stop paste" sequence', function()
+  end)
```
What should happen if "stop paste" is never received? A timeout?

justinmk commented on src/nvim/tui/input.c:
```diff
@@ -282,27 +283,44 @@ static bool handle_focus_event(TermInput *input)
+
+    if (enable && input->paste_started) {
+      // Bogus "paste start"; forward it.
+      enqueue_input(input, "<C-v><Esc>200~", sizeof("<C-v><Esc>200~") - 1);
```
@Zyx-I does this look right to you?

ZyX-I replied:
Why not simply `return false` and let the cycle handle the rest? You in any case need some kind of escaping for special characters there.

justinmk replied:
> Why not simply return false and let the cycle handle the rest?

That's what I did originally, but it did not return anything. If I am reading the logic correctly:

- With the the `rbuffer_consumed()` call on line 292, we must return something. 
- Or without it, the `RBUFFER_EACH` check on line 366 would loop back to here.

justinmk commented:
@ZyX-I Doing `flush_input` and putting the logic in `wait_input_enqueue` is definitely better. I will incorporate that.

ZyX-I commented on src/nvim/tui/input.c:
```diff
@@ -282,27 +283,45 @@ static bool handle_focus_event(TermInput *input)
 
+static void apply_pastepre(void **argv)  // MAIN thread
+{
+  apply_autocmds(EVENT_PASTEPRE, NULL, NULL, false, curbuf);
```
By the way, is it allowed to call non-API functions from TUI at all? When I was creating my example with callbacks I assumed that TUI should be considered “external” and thus only API and “utility” (i.e. rbuffer, memory functions, etc) functions should be used and thus my example constructed an Object and not a typval_T.

justinmk replied:
This is scheduled to run on the main thread, but using API is probably preferable.

I'm running into timing (or other) issues on travis, so this PR is stalled anyways.

ZyX-I replied:
When I asked about allowed I did not mean *technical* reasons not to call `apply_autocmds`. I talked about architectural issues, even if you call `vim_doautocmd(STATIC_CSTR_AS_STRING("PastePre"), STATIC_CSTR_AS_STRING("") /* group */, STATIC_CSTR_AS_STRING("") /* fname */, BOOLEAN_OBJ(false).data.boolean /* nomodeline */, &err)` (note I did not suggest using `vim_command` because I do not like the idea; also in any case every command functionality will eventually use API) this still needs to be in the main thread because here you are using a C function avoiding events dispatcher.

justinmk replied:
Ok.

Just a side note: the constant disdainful tone of your comments is exhausting. It's OK every now and then, but for it to be constant is exhausting.

bfredl commented:
@justinmk what is the time scale if this? Because something simple like [this](https://github.com/neovim/neovim/compare/master...bfredl:termfix?expand=1) could mitigate #3476 in the meanwhile (I tested with urxvt and pangoterm) if it will take some time for this to get ready.

bfredl committed:
API: nvim_put #6819

bfredl committed:
API: nvim_put #6819: try to fix Insert, Visual

wincent commented:
Any updates or progress here? Just recently started using Neovim and I'm sad that it can't be made to accept (bracketed) pastes in normal mode like Vim can.

janlazo commented:
Is this blocked indefinitely? I have https://gist.github.com/janlazo/cea7acf2711e89a9bc4f7c79612c6e54#bracketed-paste for Vim 8.0 patches for bracketed paste.

justinmk commented:
No.

justinmk committed:
paste: WIP #4448

ZyX-I committed:
TUI/paste: collect data, invoke user callback #4448

oni-link commented on src/nvim/tui/input.c:
```diff
@@ -102,17 +102,65 @@ static void tinput_done_event(void **argv)
+          || (fret.type == kObjectTypeString && fret.data.string.size)) {
+        input->paste_enabled = false;
+      }
+      api_free_object(fret);
```
Missing handling of `err` being set?

justinmk committed:
TUI/paste: push bytes directly (avoid libtermkey)

justinmk commented:
> It also appears that I understood the format of incoming string incorrectly, so additional action is needed in string_to_array.

`tk_getkeys()` translates control bytes to `<foo>` keycode strings. I don't think we want that, and it would require un-translating also (e.g. when users pastes `\t` char they want that literal tab char, not a `"<Tab>"` string).

Thus in `tinput_read_cb()` I am calling `tinput_enqueue()` directly during paste-mode (skipping `tk_getkeys()`) .

justinmk committed:
log: log_key()

justinmk committed:
lua/stdlib: cleanup

justinmk committed:
TUI/paste: define paste function as Lua builtin

- Define in Lua so that it is compiled-in (available with `-u NONE`).

TODO: Eventually we will want a 'pastefunc' option or some other way to
override the default paste handler.

justinmk committed:
TUI/paste: always flush on paste mode-change

Flush input before entering, not only when leaving, paste mode. Else
there could be pending input which will erroneously be sent to the paste
handler.

justinmk committed:
paste: abort paste if handler does not return true

justinmk committed:
paste: fixup tests

justinmk committed:
paste: use chansend() in Terminal-mode

justinmk committed:
API: nvim_put #6819

justinmk committed:
paste: use nvim_put()

bfredl commented on src/nvim/lua/vim.lua:
```diff
@@ -93,6 +93,21 @@ local function _os_proc_children(ppid)
+  -- vim.api.nvim_set_option('paste', false)
+  -- TODO: do not redraw (slow!) until paste is finished.
+  -- if eof then
+  vim.api.nvim_command('redraw')
```
should not be needed (perhaps `nvim_put` lack a call to mark buffer as changed?)

bfredl replied:
Doesn't seem so, `do_put` already calls `changed_lines` (which schedules any window containing the buffer to be redrawn already).

justinmk committed:
paste: test

justinmk committed:
paste: edge-case: handle EOL at end-of-buffer

This is "readfile()-style", see also ":help channel-lines".

justinmk committed:
API: nvim_put: Avoid "N more lines" message

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -197,26 +201,97 @@ describe('TUI', function()
-    -- Need extra time for this test, specially in ASAN.
-    screen.timeout = 60000
-    feed_command('set ruler')
+  it('paste: big burst of input', function()
```
this test and the `paste: exactly 64 bytes` test have strange failures. The screen failure shows out-of-order lines, but I think this is a display/refresh issue. Logging shows that the lines are received by `nvim_put` in the correct order.

blueyed replied:
Also locally for you?  (they work for me)

bfredl replied:
Does the buffer contents look ok afterwards?

justinmk replied:
> Also locally for you? (they work for me)

Sometimes. It's random.

> Does the buffer contents look ok afterwards?

Since it's a child Nvim can't use `helpers.expect()`. I will check by writing to a file.

justinmk replied:
Doing `PUT_CURSEND` always seems to fix it. I guess the cursor placement depends on redraws which may be skipped/deferred? And this can probably be forced `check_cursor()`? But maybe it's fine if `nvim_put` always does `PUT_CURSEND`.

bfredl replied:
Cursor position should never depend on when a redraw is done. Forcing PUT_CURSEND in normal mode breaks using `nvim_put` to replace `call setreg("x", data, mode) | normal "xp`, which I would hope it could do. Might better to have a flag for "always cursend" if this is needed for modelesslike TUI/GUI paste.

justinmk committed:
test/tui_spec: connect to child session

justinmk committed:
API: nvim_put: always PUT_CURSEND

Fixes strange behavior where sometimes the buffer contents of a series
of paste chunks (vim._paste) would be out-of-order.

Now the tui_spec.lua screen-tests are much more reliable. But they still
sometimes fail because of off-by-one cursor (caused by "typeahead race"
resulting in wrong mode; fixed later in this patch-series).

justinmk committed:
API: nvim_put: "follow" parameter

justinmk committed:
paste: phases, dots

- Send `phase` parameter to the paste handler.
- Redraw at intervals and when paste terminates.
- Show "..." throbber during paste to indicate activity.

justinmk commented:
> this test and the `paste: exactly 64 bytes` test have strange failures ... out-of-order lines.

Root cause is a [race with typeahead](https://github.com/neovim/neovim/pull/4448/commits/6b9e68bc5c866cd9ab512b018d8746a2df666a98). This "race" won't cause trouble in real-world usage, but it would be nice if we had a deterministic way to flush typeahead. (See also #6247)

Example: from Normal-mode if   "i" is followed immediately by a paste sequence:

    i<start-paste>...<stop-paste>
    ^
     "i" does not get processed in time, so the editor is in
     Normal-mode instead of Insert-mode while handling the paste.

I tried "flushing" typeahead by various means: 
- `nvim_feedkeys('','x')` 
- `exec_normal()` from `tinput_wait_enqueue()`
- `LOOP_PROCESS_EVENTS(&main_loop,…,0)` from `tinput_wait_enqueue()`

The most reliable (but not completely) workaround is to do a synchronous RPC request on the child Nvim:

```lua
    feed_data('i')
    child_session:request('nvim_eval', 1)
    feed_data('\027[200~'..expected..'\027[201~')
```

bfredl commented:
@justinmk I was thinking about a similar situation ago and though about somehow doing it the other way around: store references to events in the key input stream/typeahead buffer. I e while not concatenating all of the paste data into the typeahead buffer as is done before, `nvim_paste` would put a single "token" on the typeahead stream that would reference the paste event (or a general LuaRef to execute, in my original context). Not sure how much refactor it would incur though, probably too much for only this usecase.

bfredl commented:
A proper implementation of `nvim_mouse_input` would also requiring "embedded" data, though maybe that case is simpler, as the amount of data is small and bounded.

justinmk committed:
paste: workaround typeahead race

Workaround this failure:

    [  ERROR   ] test/functional/terminal/tui_spec.lua @ 192: TUI paste: exactly 64 bytes
    test/functional/helpers.lua:403:
    retry() attempts: 478
    test/functional/terminal/tui_spec.lua:201: Expected objects to be the same.
    Passed in:
    (table: 0x47cd77e8) {
     *[1] = 'zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz endz' }
    Expected:
    (table: 0x47cd7830) {
     *[1] = 'zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz end' }

This happens because `curwin->w_cursor.col` is sometimes decremented at
the end of `do_put`... because the editor is in Normal-mode instead of
the expected Insert-mode.

Caused by "typeahead race" (#10826): there may be queued input in the
main thread not yet processed, thus the editor mode (`State` global)
will be "wrong" during paste. Example: input "i" followed immediately by
a paste sequence:

    i<start-paste>...<stop-paste>
    ^
     "i" does not get processed in time, so the editor is in
     Normal-mode instead of Insert-mode while handling the paste.

Attempted workarounds:
- vim.api.nvim_feedkeys('','x',false) in vim._paste()
- exec_normal() in tinput_wait_enqueue()
- LOOP_PROCESS_EVENTS(&main_loop,…,0) in tinput_wait_enqueue()

ref #10826

justinmk commented:
> somehow doing it the other way around: store references to events in the key input stream/typeahead buffer ... put a single "token" on the typeahead stream

That would be very cool (see also https://github.com/neovim/neovim/pull/4448#issuecomment-196264209 ; maybe also discussed in #6247), and is generally useful for UIs/clients. This approach seems easiest. Created a ticket:  #10826

justinmk committed:
API: nvim_paste

justinmk committed:
paste: insert text "before" cursor in Insert-mode

justinmk committed:
paste: implement redo (AKA dot-repeat)

- Normal-mode redo idiom(?): prepend "i" and append ESC.
- Insert-mode only needs AppendToRedobuffLit().
- Cmdline-mode: only paste the first line.

justinmk commented:
This is essentially done except:

- redo (AKA dot-repeat) is wonky 8635e3f. Not sure what other magic spells are needed.
- cursor displays in command line after paste completes. Can hack around this by sending dummy input:
  ```
  vim.api.nvim_feedkeys('\027','n',false)
  ```

justinmk commented:
Merging after CI. Followup PR for remaining items.

justinmk committed:
paste: tickle cursor

HACK: The cursor does not get repositioned after the paste completes.
Scheduling a dummy event seems to fix it.

Test case:
0. Revert this commit.
1. Paste some text in Normal-mode.
2. Notice the cursor is still in the cmdline area.

justinmk committed:
paste: handle vim.paste() failure

- Show error only once per "paste stream".
- Drain remaining chunks until phase=3.
- Lay groundwork for "cancel".
- Constrain semantics of "cancel" to mean "client must stop"; it is
  unrelated to presence of error(s).

justinmk committed:
paste: make vim.paste() "public"

justinmk committed:
paste: handle 'nomodifiable'

- nvim_paste(): Marshal through luaeval() instead of nvim_execute_lua()
  because the latter seems to hide some errors.
- Handle 'nomodifiable' in `nvim_put()` explicitly.
- Require explicit `false` from `vim.paste()` in order to "cancel",
  otherwise assume true ("continue").

justinmk committed:
API: TRY_WRAP() for "abort-causing non-exception errors"

- Introduce TRY_WRAP() until we have an *architectural* solution.
  - TODO: bfredl idea: prepare error-handling at "top level" (nv_event).
- nvim_paste(): Revert luaeval() hack (see parent commit).
  - With TRY_WRAP() in nvim_put(), 'nomodifiable' error now correctly
    "bubbles up".

justinmk commented:
Merging after CI. [Remaining items](https://github.com/neovim/neovim/pull/4448#issue-62689821) will go in followup PR.

Marking these vim-patches as NA (cc @janlazo):

```
vim-patch:8.1.1198
vim-patch:8.1.0224
vim-patch:8.0.1299
vim-patch:8.0.0569
vim-patch:8.0.0303
vim-patch:8.0.0296
vim-patch:8.0.0244
vim-patch:8.0.0238
vim-patch:8.0.0232
vim-patch:8.0.0231
vim-patch:8.0.0230
vim-patch:8.0.0210
```

Notes:

- cursor is hidden after paste completes. Workaround: send dummy event https://github.com/neovim/neovim/pull/4448/commits/4344ac11119abd20ba911d72cf540321277dd150:
  ```
  loop_schedule_deferred(&main_loop, event_create(x_dummy_event, 0));
  ```
